### PR TITLE
[UR] Support C style arrays

### LINKED ISSUE #869: Support raw C-style array type in spec-generator
Currently the following syntax will fail to generate the spec if a c-style raw array is used in a struct type definition 
i.e.

```yml
type: size_t
name: my_array[3]
desc: [in] C-Style array

```

### PR DESCRIPTION
**Pull request by veselypeta**
This PR:
* Adds support to type members of structs as c-style arrays
* Changes `ur_exp_sampler_addr_modes_t` to use array as requested in #863 

Closes #869

### CODE REVIEW

**Review by Seanst98 — APPROVED** [review]
LGTM, thanks!

File: scripts/templates/helper.py [hunk 1]
```diff
@@ -217,6 +218,29 @@ def find_class_name(name, meta):
...
+        if not cls.is_array(name):
+            raise Exception("Cannot find array length of non-array type.")
+
+        match = re.match(cls.RE_ARRAY, name)
+        return match.groups()[1]
+    
+    @classmethod
+    def get_array_element_type(cls, name):
+        if not cls.is_array(name):
+            raise Exception("Cannot find array length of non-array type.")
```
Comment (pbalcer) [inline scripts/templates/helper.py:239]:
```suggestion
            raise Exception("Cannot find array type of non-array type.")
```

File: scripts/templates/params.hpp.mako [hunk 2]
```diff
@@ -84,6 +84,15 @@ def findMemberType(_item):
     %elif findMemberType(item) is not None and findMemberType(item)['type'] == "union":
         os << ".${iname} = ";
         ${x}_params::serializeUnion(os, ${deref}(params${access}${item['name']}), params${access}${th.param_traits.tagged_member(item)});
+    %elif th.type_traits.is_array(item['type']):
+        os << ".${iname} = [";
```
Comment (pbalcer) [inline scripts/templates/params.hpp.mako:88]:
```suggestion
        os << ".${iname} = {";
```

File: scripts/templates/params.hpp.mako [hunk 3]
```diff
@@ -84,6 +84,15 @@ def findMemberType(_item):
...
+        for(auto i = 0; i < ${th.type_traits.get_array_length(item['type'])}; i++){
+            if(i != 0){
+                os << ", ";
+            }
+            os << ${deref}(params${access}${item['name']}[i]);
+        }
+        os << "]";
```
Comment (pbalcer) [inline scripts/templates/params.hpp.mako:95]:
```suggestion
        os << "}";
```

File: scripts/templates/params.hpp.mako [hunk 4]
```diff
@@ -84,6 +84,15 @@ def findMemberType(_item):
     %elif findMemberType(item) is not None and findMemberType(item)['type'] == "union":
         os << ".${iname} = ";
         ${x}_params::serializeUnion(os, ${deref}(params${access}${item['name']}), params${access}${th.param_traits.tagged_member(item)});
+    %elif th.type_traits.is_array(item['type']):
+        os << ".${iname} = [";
+        for(auto i = 0; i < ${th.type_traits.get_array_length(item['type'])}; i++){
+            if(i != 0){
+                os << ", ";
+            }
+            os << ${deref}(params${access}${item['name']}[i]);
```
Comment (pbalcer) [inline scripts/templates/params.hpp.mako:93]:
```suggestion
<%call expr="member(iname, itype, True)">
    ${deref}(params${access}${item['name']}[i])
</%call>
```

File: include/ur.py [hunk 5]
```diff
@@ -2228,9 +2228,7 @@ class ur_exp_sampler_addr_modes_t(Structure):
         ("stype", ur_structure_type_t),                                 ## [in] type of this structure, must be
                                                                         ## ::UR_STRUCTURE_TYPE_EXP_SAMPLER_ADDR_MODES
         ("pNext", c_void_p),                                            ## [in,out][optional] pointer to extension-specific structure
-        ("addrModeX", ur_sampler_addressing_mode_t),                    ## [in] Specify the addressing mode of the x-dimension.
-        ("addrModeY", ur_sampler_addressing_mode_t),                    ## [in] Specify the addressing mode of the y-dimension.
-        ("addrModeZ", ur_sampler_addressing_mode_t)                     ## [in] Specify the addressing mode of the z-dimension.
+        ("addrModes", ur_sampler_addressing_mode_t * 3)                 ## [in] Specify the address mode of the sampler per dimension
```
Comment (kbenzie) [inline include/ur.py:2231]: Can you point me to where its documented this is how to define an array?

**Review by kbenzie — APPROVED** [review]

File: include/ur.py (hunk 5 above)
Comment (veselypeta) [reply include/ur.py:2231]: https://docs.python.org/3/library/ctypes.html#arrays

**Review by pbalcer — APPROVED** [review]

**Pull request merged**